### PR TITLE
add exponential search algorithm

### PR DESCRIPTION
**Pull request by kiarash8112**
### Describe your change:

added exponential algorithm using binary_search_recursion also tests and docs about it

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [x] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

File: searches/binary_search.py [hunk 1]
```diff
@@ -297,6 +297,49 @@ def binary_search_by_recursion(
...
+    0
+
+    >>> exponential_search([0, 5, 7, 10, 15], 15)
+    4
+
+    >>> exponential_search([0, 5, 7, 10, 15], 5)
+    1
+
+    >>> exponential_search([0, 5, 7, 10, 15], 6)
+
```
Comment (cclauss) [inline searches/binary_search.py:325]:
```suggestion
    >>> exponential_search([0, 5, 7, 10, 15], 0)
    0
    >>> exponential_search([0, 5, 7, 10, 15], 15)
    4
    >>> exponential_search([0, 5, 7, 10, 15], 5)
    1
    >>> exponential_search([0, 5, 7, 10, 15], 6) = None
    True
```

File: searches/binary_search.py [hunk 2]
```diff
@@ -297,6 +297,49 @@ def binary_search_by_recursion(
         return binary_search_by_recursion(sorted_collection, item, midpoint + 1, right)
 
 
+def exponential_search(sorted_collection: list[int], item: int) -> int | None:
```
Comment (cclauss) [inline searches/binary_search.py:300]: I am not a fan of functions that return multiple datatypes.  If the search fails, let's return `-1` (like `"Hello".find("Z")`) or raise a `ValueError` (like `"Hello".index("Z")`).
```suggestion
def exponential_search(sorted_collection: list[int], item: int) -> int:
```

File: searches/binary_search.py [hunk 3]
```diff
@@ -297,6 +297,49 @@ def binary_search_by_recursion(
...
+    )
+
+
+if __name__ == "__main__":
+    import doctest
+
+    doctest.testmod()
+
+
 if __name__ == "__main__":
```
Comment (cclauss) [inline searches/binary_search.py:343]: Only one of these.  Run the tests first and let the user interact afterward.
```suggestion
```

**Review by cclauss — CHANGES_REQUESTED** [review]
Nice addition!

For optional extra credit, add a `timeit` benchmark to show relative performance.

File: searches/binary_search.py [hunk 4]
```diff
@@ -297,7 +297,46 @@ def binary_search_by_recursion(
...
+    >>> exponential_search([0, 5, 7, 10, 15], 5)
+    1
+    >>> exponential_search([0, 5, 7, 10, 15], 6) = None
+    True
+    """
+    sorted_collection = sorted(sorted_collection)
+    bound = 1
+    while bound < len(sorted_collection) and sorted_collection[bound] < item:
+        bound *= 2
+    left = int(bound / 2)
```
Comment (cclauss) [inline searches/binary_search.py:328]:
```suggestion
    left = bound // 2
```

File: searches/binary_search.py [hunk 5]
```diff
@@ -297,7 +297,46 @@ def binary_search_by_recursion(
...
+
+    Examples:
+    >>> exponential_search([0, 5, 7, 10, 15], 0)
+    0
+    >>> exponential_search([0, 5, 7, 10, 15], 15)
+    4
+    >>> exponential_search([0, 5, 7, 10, 15], 5)
+    1
+    >>> exponential_search([0, 5, 7, 10, 15], 6) = None
+    True
```
Comment (cclauss) [inline searches/binary_search.py:322]:
```suggestion
    >>> exponential_search([0, 5, 7, 10, 15], 6)
    -1
```

**Comment by cclauss** [comment]
I thought more about sorting the sorted collection… it is a bad idea.  The second item in might be different in the two lists so the function might tell me that is the second item when in my list it is the third item.  So, the implementations should raise ValueError on unsorted input.

**Review by cclauss — APPROVED** [review]
This is a really slick addition.  Thanks for doing it!

**Comment by kiarash8112** [comment]
@cclauss  thank you for your suggestions :)

**Comment by tianyizheng02** [comment]
@cclauss Since exponential search is a different algorithm from binary search, why was it placed in the `binary_search.py` file? Shouldn't it have gotten its own file?

File: searches/binary_search.py [hunk 6]
```diff
@@ -297,12 +282,78 @@ def binary_search_by_recursion(
...
+    -1
+    """
+    if list(sorted_collection) != sorted(sorted_collection):
+        raise ValueError("sorted_collection must be sorted in ascending order")
+    bound = 1
+    while bound < len(sorted_collection) and sorted_collection[bound] < item:
+        bound *= 2
+    left = bound // 2
+    right = min(bound, len(sorted_collection) - 1)
+    last_result = binary_search_by_recursion(
```
Comment (cclauss) [inline searches/binary_search.py:316]: @tianyizheng02 `exponential_search()` embeds a call to `binary_search_by_recursion()` but as the benchmarks demonstrate delivers a faster result.  Looking at the first commit makes this clearer.

**Pull request merged**